File: leetcode/bus_routes/bus_routes.cpp
/*
 * Time Complexity: O(N * R)
 *   - N is the total number of stops across all routes.
 *   - R is the number of routes.
 *   - Each stop and bus route may be visited once in the BFS.
 *
 * Space Complexity: O(N + R)
 *   - For the busByStops map, visited sets, and BFS queue.
 */

#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <queue>
#include <utility>

using std::vector;
using std::unordered_set;
using std::unordered_map;
using std::queue;
using std::pair;

class Solution {
public:
    int numBusesToDestination(vector<vector<int>>& routes, int source, int target) {
        unordered_map<int, vector<int>> busByStops;
        for(int i = 0; i < routes.size(); i++) {
            for(int j = 0; j < routes[i].size(); j++) {
                int stop = routes[i][j];
                busByStops[stop].push_back(i);
            }
        }

        queue<pair<int,int>> q;
        q.push({source, 0});
        unordered_set<int> visitedBus;
        unordered_set<int> visitedStops;

        while(!q.empty()) {
            int currStop = q.front().first;
            int currDist = q.front().second;

            if(currStop == target) return currDist;

            q.pop();

            visitedStops.insert(currStop);
            for(int bus : busByStops[currStop]) {
                if(visitedBus.find(bus) != visitedBus.end()) continue;

                visitedBus.insert(bus);

                for(int stop : routes[bus]) {
                    if(visitedStops.find(stop) == visitedStops.end()) {
                        q.push({stop, currDist+1});
                    }
                }
            }
        }

        return -1;
    }
};

int main() {
    // create the following test case: [[0,1,6,16,22,23],[14,15,24,32],[4,10,12,20,24,28,33],[1,10,11,19,27,33],[11,23,25,28],[15,20,21,23,29],[29]]
    vector<vector<int>> routes = {
        {0, 1, 6, 16, 22, 23},
        {14, 15, 24, 32},
        {4, 10, 12, 20, 24, 28, 33},
        {1, 10, 11, 19, 27, 33},
        {11, 23, 25, 28},
        {15, 20, 21, 23, 29},
        {29}
    };
    int source = 4;
    int target = 21;
    Solution solution;
    int result = solution.numBusesToDestination(routes, source, target);
    // Expected output: 2
    std::cout << "Number of buses to destination: " << result << std::endl;
    return 0;
}

File: leetcode/bus_routes/bus_routes.java
package bus_routes;

import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.Set;

class Solution {
    public int numBusesToDestination(int[][] routes, int source, int target) {
        if (source == target) return 0;

        int n = routes.length;
        if (n == 0) return -1;

        Map<Integer, List<Integer>> busByStops = new HashMap<>();
        for (int i = 0; i < n; i++) {
            for (int stop : routes[i]) {
                busByStops.computeIfAbsent(stop, k -> new ArrayList<>()).add(i);
            }
        }

        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{source, 0}); // {stop, distance}
        
        Set<Integer> visitedBuses = new HashSet<>();
        Set<Integer> visitedStops = new HashSet<>();
        
        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            int currStop = current[0];
            int currDist = current[1];

            if (currStop == target) return currDist;

            visitedStops.add(currStop);
            for (int bus : busByStops.getOrDefault(currStop, new ArrayList<>())) {
                if (visitedBuses.contains(bus)) continue;
                visitedBuses.add(bus);

                for (int stop : routes[bus]) {
                    if (!visitedStops.contains(stop)) {
                        queue.offer(new int[]{stop, currDist + 1});
                    }
                }
            }
        }

        return -1;
    }
    
}
